### PR TITLE
Extract listeners to own classes

### LINKED ISSUE #114: Refactor listeners into separate classes
Currently, we have a lot of anonymous classes acting as listeners for our UI elements. Let's refactor our code so we extract these anonymous listener classes into separate listener classes that can be unit tested.

### PR DESCRIPTION
**Pull request by kaden-dreger**
In this PR I extracted out common listeners into their own classes such that it reduces repeated code and can be easily unit tested. I have added unit tests for those extracted listeners as well. This closes #114.

### CODE REVIEW

File: app/src/androidTest/java/com/example/habitsmasher/MainActivityTest.java
Comment (kaden-dreger) [inline app/src/androidTest/java/com/example/habitsmasher/MainActivityTest.java:1221]: This was changed to check for the profile fragment rather than just wait an arbitrary time.

File: app/src/main/java/com/example/habitsmasher/listeners/SuccessListenerWithToast.java
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/listeners/SuccessListenerWithToast.java:13]: If you choose to combine the FailureListener into one class with the boolean flag passed in as a parameter, I'd do the same to this one. Just another design decision to consider.

File: app/src/main/java/com/example/habitsmasher/listeners/SuccessListenerWithToast.java
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/listeners/SuccessListenerWithToast.java:32]: - [ ] add logging

File: app/src/main/java/com/example/habitsmasher/ui/dashboard/AddHabitDialog.java
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/ui/dashboard/AddHabitDialog.java:154]: Wow! I like how much simpler it is, lots of lines of code saved.

File: app/src/test/java/com/example/habitsmasher/test/ListenersTest.java
Comment (rudy-patel) [inline app/src/test/java/com/example/habitsmasher/test/ListenersTest.java:38]: I would prefer if the tests corresponding to **each class** were split up into their own files. Even if it's just one/two tests per file, it just really helps when you need to look for the tests for **that specific class**. This shouldn't be too bad, and makes the test class a lot easier to follow and modify in the future.
- [ ] move unit tests to their own test class for each listener

File: app/src/test/java/com/example/habitsmasher/test/ListenersTest.java
Comment (rudy-patel) [inline app/src/test/java/com/example/habitsmasher/test/ListenersTest.java:69]: Lots of mocking, I like it!

**Review by rudy-patel — CHANGES_REQUESTED** [review]
Some good changes in here, thanks for doing this. It's a lot cleaner. I've left some tasks that I'd like to see completed before we merge this in. After they are completed, you have my approval.

File: app/src/test/java/com/example/habitsmasher/test/ListenersTest.java
Comment (rudy-patel) [inline app/src/test/java/com/example/habitsmasher/test/ListenersTest.java:98]: Can we be more specific about the **type** of exception we expeect here?

File: app/src/test/java/com/example/habitsmasher/test/ListenersTest.java
Comment (rudy-patel) [inline app/src/test/java/com/example/habitsmasher/test/ListenersTest.java:77]: If I'm reading this correctly, the `verify()` method only ensures that the method was **called X amount of times**, not on the correctness of the result.

Do we also need to assert on some expected condition, or is the fact that we received no errors good enough?

File: app/src/main/java/com/example/habitsmasher/listeners/SwipeListener.java
Comment (jaskim9824) [inline app/src/main/java/com/example/habitsmasher/listeners/SwipeListener.java:16]: From what I can see, this listener class isn't used by HabitListFragment, which uses a listener of this sort. We should probably integrate this class into HabitListFragment.

- [ ] Use SwipeListener in HabitListFragment

File: app/src/main/java/com/example/habitsmasher/listeners/SwipeListener.java
Comment (jaskim9824) [inline app/src/main/java/com/example/habitsmasher/listeners/SwipeListener.java:44]: Since we already pass in an instance of HabitListFragment into this listener, we should probably just call the methods openEditDialogBox() and updateListAfterDelete() in the HabitListFragment class here instead of repeating the same logic.

- [ ] Replace logic with calls to method in HabitListFragment

Comment (jaskim9824) [reply app/src/main/java/com/example/habitsmasher/listeners/SwipeListener.java:44]: This'll reduce the number of fields you'll have to pass in also, as the user, habit list and adapter are used in the methods defined in HabitListFragment.

File: app/src/main/java/com/example/habitsmasher/listeners/SwipeListener.java
Comment (jaskim9824) [inline app/src/main/java/com/example/habitsmasher/listeners/SwipeListener.java:15]: Also, are we going to use this class for the swipe options in HabitEventListFragment or just create a new SwipeListener class to use in there? I know that this hasn't been implemented in develop yet but I rather we have an idea of how that is going to be handled.

**Review by jaskim9824 — CHANGES_REQUESTED** [review]
Looks good overall, definitely a lot cleaner and more organized. Left some changes I would like see done before I give this my approval.

File: app/src/main/java/com/example/habitsmasher/listeners/SwipeListener.java
Comment (jaskim9824) [inline app/src/main/java/com/example/habitsmasher/listeners/SwipeListener.java:61]: These methods are already implemented in the HabitListFragment class. Therefore, instead of creating a new copy of the methods here, we can just call _fragment.openEditDialogBox() and _fragment.updateListAfterDelete() in the swipe options clicked method here. This reduces the need to pass in the habit list, the adapter, and the fragment manager into the listener. Apologies for not making this clear in my previous review.

**Review by jaskim9824 — CHANGES_REQUESTED** [review]

File: app/src/test/java/com/example/habitsmasher/test/ListenersTest.java
Comment (rudy-patel) [reply app/src/test/java/com/example/habitsmasher/test/ListenersTest.java]: Any updates on this? 

I see we are throwing a `NullPointerException`, so we can probably be more specific on the **type** of Exception we expect. If literally **anything** goes wrong, this test will still pass, because we are expecting **any Exception**. The same thing applies to any other test that does this.

- [ ] change Exception to NullPointerException

File: app/src/test/java/com/example/habitsmasher/listeners/FailureListenerTest.java
Comment (rudy-patel) [inline app/src/test/java/com/example/habitsmasher/listeners/FailureListenerTest.java:36]: The last little thing here, can we be more specific on the **type** of exception we expect?

This test will pass if literally **anything** goes wrong in the code. Your listener classes throw a `NullPointerException`, so that is probably what we should be expecting here instead of just `Exception`. The same thing applies to any other tests that do this.
- [ ] change Exception to NullPointerException

**Review by rudy-patel — CHANGES_REQUESTED** [review]
Just one last little thing to beef up the tests, but after that is dealt with, I'm cool with this. See the comment/task I left.

**Review by jaskim9824 — APPROVED** [review]

**Review by jaskim9824 — APPROVED** [review]

File: app/src/main/java/com/example/habitsmasher/listeners/ClickListenerForDaysOfTheWeek.java
Comment (Nova-Drive) [inline app/src/main/java/com/example/habitsmasher/listeners/ClickListenerForDaysOfTheWeek.java:18]: Thank you for this!

File: app/src/main/java/com/example/habitsmasher/ui/dashboard/AddHabitDialog.java
Comment (Nova-Drive) [reply app/src/main/java/com/example/habitsmasher/ui/dashboard/AddHabitDialog.java:165]: I agree!

**Review by Nova-Drive — APPROVED** [review]
Looks great! I never knew you could unit test stuff like that. Everything looks much cleaner now.

**Review by rudy-patel — APPROVED** [review]

**Pull request merged**
